File: 345.Reverse Vowels of a String.py
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        vowels = set(['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'])
        i, j = 0, len(s) - 1
        string = list(s)
        while i < j:
            if string[i] in vowels and string[j] in vowels:
                string[i], string[j] = string[j], string[i]
            elif string[i] in vowels:
                i -= 1
            elif string[j] in vowels:
                j += 1
            i += 1
            j -= 1
        return ''.join(string)

File: 345.Reverse Vowels of a String.cpp
class Solution {
public:
    string reverseVowels(string s) {
        vector<int> vowels;
        
        for (int i=0; i<s.length(); i++) {
            if (s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u'
                || s[i] == 'A' || s[i] == 'E' || s[i] == 'I' || s[i] == 'O' || s[i] == 'U') {
                    vowels.push_back(i);
            }
        }
        
        int size = vowels.size();
        for (int i=0; i<size/2; i++) {
            char tmp = s[vowels[size-1-i]];
            s[vowels[size-1-i]] = s[vowels[i]];
            s[vowels[i]] = tmp;
        }
        
        return s;
    }
};

//====================== improve version =========================

class Solution {
public:
    string reverseVowels(string s) {
        int i = 0, j = s.size() - 1;
        while (i < j) {
            i = s.find_first_of("aeiouAEIOU", i);
            j = s.find_last_of("aeiouAEIOU", j);
            if (i < j) {
                swap(s[i++], s[j--]);
            }
        }
        return s;
    }
};
